rafthttp/snapshot_hub.go

```diff
@@ -0,0 +1,170 @@
+	return &snapshotHub{
+		tr:         tr,
+		urlPickers: urlPickers,
+		r:          r,
+		snapst:     snapst,
+	}
+}
+
+// registerSnap registers snapshot in snapshot hub.
+// The registered snapshot waits to be written out through writeSnapTo func.
+// If it exceeds wait timeout, the snapshot is closed and deregistered.
+// At most one snapshot could be registered at the same time, or it panics.
+func (s *snapshotHub) registerSnap(to types.ID, index uint64) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.hasSnap {
+		plog.Panicf("unexpected snapshot register when previous one has not finished")
+	}
+	s.hasSnap = true
+	s.snapTo = to
+	s.snapIndex = index
+	s.snapClose = time.AfterFunc(snapshotWaitTimeout, func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if !s.hasSnap {
+			return
+		}
+		s.hasSnap = false
+		// close the snapshot in the store
+		s.snapst.WriteSnapTo(&errWriter{errors.New("cancelled")})
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out due to timeout", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(to), raft.SnapshotFailure)
+	})
+}
+
+// writeSnapTo writes the snapshot with the given index into the given writer.
+func (s *snapshotHub) writeSnapTo(w io.Writer, index uint64) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.hasSnap {
+		return fmt.Errorf("failed to find available snapshot")
+	}
+	if s.snapIndex != index {
+		return fmt.Errorf("failed to find wanted snapshot")
+	}
+	s.hasSnap = false
+	s.snapClose.Stop()
+
+	err := s.snapst.WriteSnapTo(w)
+	if err != nil {
+		plog.Infof("snapshot [index: %d, to: %s] failed to be sent out (%v)", index, s.snapTo, err)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFailure)
+	} else {
+		plog.Infof("snapshot [index: %d, to: %s] succeeded to be sent out", index, s.snapTo)
+		s.r.ReportSnapshot(uint64(s.snapTo), raft.SnapshotFinish)
+	}
+	return err
+}
+
+// processSnap processes the given raft snapshot message.
+// First, it sends requests to get full snapshot for raft snapshot message. Then
+// it lets snapshot store read out the snapshot. After the full snapshot is
+// received, it lets raft process the raft snapshot message.
+func (s *snapshotHub) processSnap(m raftpb.Message) {
+	remote := types.ID(m.From)
+	index := m.Snapshot.Metadata.Index
+	snapr, err := s.requestSnap(remote, index)
+	if err != nil {
+		plog.Warningf("failed to request snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	defer snapr.Close()
+
+	if err := s.snapst.ReadSnapFrom(snapr, index); err != nil {
+		plog.Warningf("failed to read snapshot [index: %d] (%v)", m.Snapshot.Metadata.Index, err)
+		return
+	}
+	if err := s.r.Process(context.TODO(), m); err != nil {
+		plog.Warningf("failed to process raft message (%v)", err)
+	}
+}
+
+func (s *snapshotHub) requestSnap(remote types.ID, index uint64) (io.ReadCloser, error) {
+	up, ok := s.urlPickers[remote]
+	if !ok {
+		return nil, fmt.Errorf("cannot find url picker")
```

Comment:
what do you mean by cannot find a picker? this will be logged out as warning. I do not expect our user to understand the internal of rafthttp.